### PR TITLE
Add Sp19 METEO473 tutorials

### PR DESCRIPTION
**Pull request by ShawnMurd**
These are the tutorials I created back in Spring 2019 for METEO 473. The tutorials are all regular .py files, so we'll probably want to convert them into jupyter notebooks before we use them in the club. I think these tutorials would be a great starting point for getting students spun up with the basics of Python and also expose them to some basic plotting and data wrangling with pandas and xarray.

### CODE REVIEW

File: tutorials/473_py_tutorials/tutorial1_imports_datatypes.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+print()
+
+
+#---------------------------------------------------------------------------------------------------
+# Python Dictionary Datatypes
+#---------------------------------------------------------------------------------------------------
+
+# Dictionaries are similar to databases where each "key" is matched with a "value". Similarly to
+# lists, "keys" and "values" can be any kind of Python object
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial1_imports_datatypes.py:181]: DIctionary values can by any kind of python object.  You can't use an object as a dictionary key if you can change it, so lists and dictionaries are out.

File: tutorials/473_py_tutorials/tutorial1_imports_datatypes.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+# Comments are chunks of text within a computer program that tell anyone reading the program what
+# is going on. Thus, comments are for the human user, not the computer. It is considered good
+# computer programming practice to include comments in you code so others can follow what you are
+# doing.
+
+# There are two ways of writing comments in Python:
+#     Starting a line with a # makes that line a comment
+#     Any lines between triple quotes (''' or """) are comments (this is referred to as a block 
+#         comment)
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial1_imports_datatypes.py:22]: Python comments start with `#` and continue to the end of the line: Python ignores everything between the `#` and the newline.

Things between triple quotes are string literals (like those between normal quotes `"like this"` or `'this'` but able to include multiple lines).  If you don't do anything with the literal, like assign it to a variable, what happens depends on how you run the code.  If it's in a script, it just gets ignored.  If it's in a python interpreter, it will print the string to your screen, often in a less-readable form (IPython and Jupyter only print the last thing in a cell, so this comes up less often there).  The `help()` function couldn't access docstrings if they were comments, but that's a weaker argument because most other languages provide help using comments.

In short: surrounding a block of code with triple quotes usually has the same effect as the "Comment/Uncomment Code" button in most IDEs (and can be useful for debugging), but I'm hesitant to call it a comment because it can be a good way to get a large block of text into a variable.

File: tutorials/473_py_tutorials/tutorial1_imports_datatypes.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+print(i.itemsize)
+
+j = np.int16(1)
+print(j.itemsize)
+
+x = np.float16(1)
+print(x.itemsize)
+
+y = np.float64(1)
+print(y.itemsize)
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial1_imports_datatypes.py:211]: The smaller size means you can fit more numbers in a given amount of memory, but does mean you can't fit as much information in them.  For example, 8-bit integers can store numbers roughly -100 &ndash; 100, while 64-bit numbers can go up over a quintillion in each direction.  For floating-point numbers, the difference is in the number of significant figures carried through a calculation, and the allowed exponents in scientific notation.

```python
>>> np.iinfo(np.int8)
iinfo(min=-128, max=127, dtype=int8)
>>> np.iinfo(np.int64)
iinfo(min=-9223372036854775808, max=9223372036854775807, dtype=int64)
>>> np.finfo(np.float16)
finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)
>>> np.finfo(np.float128)
finfo(resolution=1.0000000000000000007e-18, min=-1.189731495357231765e+4932, max=1.189731495357231765e+4932, dtype=float128)
```

File: tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+x = 2.1
+y = 9.0
+i = int(3)
+j = int(2)
+A = np.array([2.0, 3.0, 5.0])
+B = np.array([10.1, 9.9, 3.0])
+
+# Rules of Thumb:
+#     An operation between two integers yields an integer (with the exception of division, integer
+#         division yields a float)
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py:30]: Perhaps write "integer division" as "division of two integers" or "quotient of two integers" to avoid confusion with the floor division operator `//`?

File: tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+print(y / j)
+
+# Exponents are performed using **
+
+print(y ** 2)
+print(A ** 3)
+
+# The remainder from a division operation can be returned using %
+
+print(i % j)
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py:51]:
```suggestion
print(i % j)

# The quotient from integer division is returned by //
print (i // j)
print(i // j + i % j)
```
Perhaps the only place this would be used is pre-allocating an array to hold the result of slicing:
```python
arr = np.eye(i)
tmp = np.empty(i // j)
tmp[:, :] = arr[::j, ::j]
```
but the only place I can see that being useful is if you want to put the even rows of a table after all the odd rows, so probably this is irrelevant here.

File: tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+# More complex operations can be performed using numpy
+# Note that "np.log" refers to the natural logarithm (i.e. ln)
+
+print(np.cos(np.pi / 4))
+print(np.exp(10))
+print(np.arcsin(0.5))
+print(np.log(np.e ** 2))
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py:59]:
```suggestion
print(np.cos(np.pi / 4))
print(np.exp(10))
print(np.arcsin(0.5))
print(np.log(np.e ** 2))

# The advantage of using NumPy is that it can operate on whole arrays at once:
print(np.exp(A))
```
Well, advantage over the stdlib `math` and `cmath` modules.

File: tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+if (x > 3) and (y > 3):
+    print('Both numbers are greater than 3')
+else:
+    print('One or both of these numbers are less than 3')
+
+if (x > 3) or (y > 3):
+    print('One of these numbers is greater than 3')
+else:
+    print('Neither of these numbers are less than 3')
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py:129]:
```suggestion
    print('Both of these numbers are less than 3')
```

File: tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    print(a)
+
+x = 1
+
+print('Basic while loop')
+while x <= 5:
+    x = x + 2
+    print(x)
+
+# The "range" function can be used to loop through a sequence of numbers
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial2_operations_conditionals_loops.py:154]:
```suggestion
# The "range" function can be used to produce a sequence of numbers to loop through 
```

File: tutorials/473_py_tutorials/tutorial3_str_formatting_io.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+string = '%d is %s than %.3f'
+
+ints = [2, 9, 6, 7]
+floats = [3.2, 6.8, 1.2, 9.4]
+
+for i, f in zip(ints, floats):
+    if i > f:
+        print(string % (i, 'greater', f))
+    elif i < f:
+        print(string % (i, 'less', f))
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial3_str_formatting_io.py:34]: There's a few other ways, but this is probably the most common (available in C and Matlab).  The full list, pulled from [the tutorial](https://docs.python.org/3/tutorial/inputoutput.html#fancier-output-formatting):

- %-formatting, available in python 2, C, and Matlab: `"%.3f" % (num,)`
- The format method, available in python 2.7 and 3+ and maybe C++: `"{:.3f}".format(num)`
- f-strings, available since python 3.6: `f"{num:.3f}"`
- [Template strings](https://docs.python.org/3/library/string.html#template-strings), available for a while (I've never seen them used): `string.Template("$num").substitude(num=num)`

File: tutorials/473_py_tutorials/tutorial3_str_formatting_io.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+for l in contents:
+    print(l.strip())
+
+fptr.close()
+
+# Parting thoughts about file I/O:
+#
+# The "open" function works well for basic text files, but in meteorology we often deal with large
+# files that are in other formats (e.g. csv, xls, netcdf, grib). These formats are not handled by
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial3_str_formatting_io.py:107]: `open` can handle CSV just fine, it's just that working with the result can be tricky.  The stdlib `csv` module deals with parsing the data into fields properly, but the pandas functions make the result much easier to work with.

You could also, technically, handle xls, netCDF, and grib files with `open(name, "rb")`, but you'd need to write many hundreds of lines of code based on documentation written by people who assume the reader knows most of this stuff already before getting to the actual data in the file in a useful format.

File: tutorials/473_py_tutorials/tutorial3_str_formatting_io.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+fptr.seek(0)
+
+# Read all the lines of fptr into a list
+
+contents = fptr.readlines()
+
+print("This file has %d lines" % len(contents))
+
+# The strip function can be used to remove \n or \t
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial3_str_formatting_io.py:97]:
```suggestion
# The strip function can be used to remove \n or \t from the ends of the string
```

File: tutorials/473_py_tutorials/tutorial3_str_formatting_io.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+for l in contents:
+    print(l.strip())
+
+fptr.close()
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial3_str_formatting_io.py:102]:
```suggestion
fptr.close()

# You can use a `with` statement to make sure the file gets closed
with open("test_file.txt", "r") as fptr:
    for line in fptr:
        print(line.strip())

# Python will call fptr.close() when the indented block ends
```
This is useful if you're doing a lot with the build-in open, less so if using other functions.

File: tutorials/473_py_tutorials/tutorial4_1_def_fct.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+def right_triangle(a, b):
+    """
+    Returns the hypotenuse of a right triangle given the lengths of the two legs.
+    Inputs:
+        a = Length of the first leg
+        b = Length of the second leg
+    Outputs:
+        c = Length of the hypotenuse
+    """
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial4_1_def_fct.py:34]: If you turn this into a notebook, you should show how to access this help:
```python
help(right_triangle)  # All python interpreters
right_triangle?  # IPython interpreters
right_triangle??  #IPython interpreters
```

File: tutorials/473_py_tutorials/tutorial6_pandas.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+#     RH = Relative humidity in %
+#     p = Pressure in mb
+#     dir = Direction wind is blowing to, in degrees from north
+#     spd = Wind speed in m/s
+#     qc = Quality control flag (1 = problem with wind data, 0 = no problem)
+
+# mm_data is a csv file. CSV stands for "comma-separated-values" and is a type of text file that 
+# has data stored in an orderly fashion, often separated by commas or spaces. The file mm_data is 
+# an example of a CSV. Feel free to open this file using a text editor in order to examine the 
+# structure of the file.
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial6_pandas.py:39]: The version here really looks like a fixed-width file, rather than a comma-delimeted one.  Did you mean to commit a different version of the file?

File: tutorials/473_py_tutorials/tutorial6_pandas.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+print(mm_df['T'])
+print()
+
+print(mm_df['RH'][:10])
+print()
+
+# We can also only select rows where a certain condition is true using the loc attribute. Here we 
+# only select the rows where qc is set to 0.
+
+print(mm_df.loc[mm_df['qc'] == 0])
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial6_pandas.py:75]: This seems like it would be better suited for [the `.query()`](https://pandas.pydata.org/docs/user_guide/indexing.html#the-query-method) or [the `.where()`](https://pandas.pydata.org/docs/user_guide/indexing.html#the-where-method-and-masking) methods than the `.loc` accessor:
```python
print(mm_df.query("qc == 0"))
```
[`.loc` is designed for using the labels for the rows and columns](https://pandas.pydata.org/docs/user_guide/10min.html#selection): since you didn't set row labels, I'm not sure how useful it is to demonstrate that type of indexing here.

File: tutorials/473_py_tutorials/tutorial6_pandas.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+# Adding New Series to a DataFrame
+#---------------------------------------------------------------------------------------------------
+
+# New series (columns) can be added to a Pandas DataFrame. For example, we might want another series
+# for potential temperature (theta) in Kelvin. We can define this new series using the existing
+# temperature and pressure series
+
+mm_df['theta'] = (mm_df['T'] + 273.15) * ((1000.0 / mm_df['p']) ** (287.04 / 1005.0))
+
+print(mm_df[:10])
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial6_pandas.py:103]:
```suggestion
print(mm_df.head(10))
```
Perhaps not necessary here, but it might avoid weird interactions with the default indexer.

File: tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+# indices corresponding to positive latitudes (negative latitudes occur in the southern 
+# hemisphere!)
+
+print(sst_file.variables['lat'][:])
+
+# We can use the np.where function to find the indices where the latitude values are positive
+
+lat_inds = np.where(sst_file.variables['lat'][:] > 0)[0]
+lats = sst_file.variables['lat'][lat_inds]
+print(lats)
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py:83]: Not using boolean indexing?

File: tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+#---------------------------------------------------------------------------------------------------
+
+# Now that we have a 2D array of sea surface temperatures, let's plot these values onto a map. To
+# do this, we will use the CartoPy module  
+
+# First, we need to set up the figure and add a single axes to the figure. In the past, we used the
+# function plt.subplots to create several axes on one figure (each subplot is an axes object in
+# Python). Now, we will use the fig.add_subplot method to add a single axes to the figure. The
+# reason for doing this is because CartoPy only works with axes objects in Python, not figure 
+# objects.
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py:113]: You can also use the `subplot_kw` to pass arguments on to the `fig.subplots` calls.  E.g.:
```python
fig, ax = plt.subplots(subplot_kw={"projection": ccrs.PlateCarree()})
```
Fun fact: the similar XArray keyword argument is spelled `subplot_kws`, so I tend to forget which to use where.

File: tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+# Python). Now, we will use the fig.add_subplot method to add a single axes to the figure. The
+# reason for doing this is because CartoPy only works with axes objects in Python, not figure 
+# objects. 
+
+# In the fig.add_subplot method, the first argument is the the same as nrows in plt.subplots, the
+# second argument is ncols (also the same as plt.subplots), and the third argument is the axes
+# number we are working with (in this case, we only have one axes). The projection keyword argument
+# specifies the map projection we wish to plot our data on. Other map projections offered by 
+# CartoPy can be found here:
+# https://scitools.org.uk/cartopy/docs/v0.15/crs/projections.html#cartopy-projections
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py:120]: All Cartopy projections expect x and y to be specified in projected coordinates, that is, with numbers you can put on the map directly with a ruler.  For the default Plate-Carree projection, the projection happens to be the same as the latitude and longitude.  For all other projections, including the Plate-Carree projection centered on a longitude of 90°W, the coordinates on the map and the coordinates on the data are different, and you need to use the `transform` argument to the plotting functions to tell Matplotlib how they're related.

If you really want them using the other projections, you should probably have a second example showing how to show the data in a Mercator, Lambert Conformal, or Geostationary projection without having the interesting bits squished into a square kilometer or so somewhere in the middle of the map.

File: tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+# respectively) that are the same shape as the sea surface temperature array. Currently, lats and 
+# lons are 1D whereas sst is 2D. To make lats and lons 2D, we will use the np.meshgrid function. 
+
+lons, lats = np.meshgrid(lons, lats)
+
+# Now we can plot sea surface temperatures using the pcolormesh function. pcolormesh creates a 
+# filled contour plot, and Python has many colormaps that can be used to color this filled
+# contour plot. We can specify the colormap we wish to use with the cmap keyword. Other available
+# colormaps can be found here:
+# https://matplotlib.org/examples/color/colormaps_reference.html
```
Comment (DWesl) [inline tutorials/473_py_tutorials/tutorial7_netcdf_2d_plots.py:136]: Matplotlib has a guide to choosing colormaps here: https://matplotlib.org/stable/tutorials/colors/colormaps.html.
It mentions some of the things to consider when choosing colormaps and some of the problems that might come up.
There's also https://colorbrewer2.org and https://seaborn.pydata.org/tutorial/color_palettes.html, though Seaborn's designed for Pandas dataframes more than for the multidimensional data usually used in meteorology.

**Review by DWesl — COMMENTED** [review]
Looks good overall.  I have a few comments in the files.

**Comment by ShawnMurd** [comment]
Thanks for the comments! Stupid question: As I make the changes you recommended to `ShawnMurd/wxdatasci:master`, will that update the PR? Or do I need to submit a new PR after making the changes?

**Comment by DWesl** [comment]
Pushing to `ShawnMurd/wxdatasci:master` will update this PR.  

If you want to take the (unnecessarily) long way of doing this, you could create a new branch, make the changes, push the new branch to GitHub, open a PR to merge your work into `ShawnMurd/wxdatasci:master` (the branch this PR is based on), then merge the new PR to update this PR, but I can't think of a reason to do that for a PR that touches fewer than many thousands of lines of code.
